File: training-data/go/longest_palindromic_subsequence.go
package main

import "fmt"

func lps(s string) int {
	n := len(s)
	if n == 0 {
		return 0
	}
	dp := make([]int, n)
	for i := n - 1; i >= 0; i-- {
		dp[i] = 1
		prev := 0 // dp[i+1][j-1] from the previous row
		for j := i + 1; j < n; j++ {
			tmp := dp[j]
			if s[i] == s[j] {
				dp[j] = prev + 2
			} else {
				dp[j] = max(dp[j], dp[j-1])
			}
			prev = tmp
		}
	}
	return dp[n-1]
}

func main() {
	for _, s := range []string{"bbbab", "character", "agbdba", "racecar", "cbbd"} {
		fmt.Println(s, lps(s), "min insertions:", len(s)-lps(s))
	}
}

File: training-data/csharp/LongestPalindromicSubsequence.cs
using System;

class LongestPalindromicSubsequence
{
    static int Lps(string s)
    {
        int n = s.Length;
        if (n == 0) return 0;
        var dp = new int[n];
        for (int i = n - 1; i >= 0; i--)
        {
            dp[i] = 1;
            int prev = 0;
            for (int j = i + 1; j < n; j++)
            {
                int tmp = dp[j];
                dp[j] = s[i] == s[j] ? prev + 2 : Math.Max(dp[j], dp[j - 1]);
                prev = tmp;
            }
        }
        return dp[n - 1];
    }

    static void Main()
    {
        foreach (var s in new[] { "bbbab", "character", "agbdba", "racecar", "cbbd" })
            Console.WriteLine($"{s} {Lps(s)} (insertions to palindrome: {s.Length - Lps(s)})");
    }
}
